examples/amm/src/contract.rs

```diff
@@ -0,0 +1,206 @@
+        _context: &OperationContext,
+        _argument: (),
+    ) -> Result<ExecutionResult<Self::Message>, AmmError> {
+        self.balance0.set(0);
+        self.balance1.set(0);
+        self.total_shares.set(0);
+        Ok(ExecutionResult::default())
+    }
+
+    async fn execute_operation(
+        &mut self,
+        _context: &OperationContext,
+        operation: Self::Operation,
+    ) -> Result<ExecutionResult<Self::Message>, AmmError> {
+        match operation {
+            Self::Operation::Swap {
+                input_token_idx,
+                output_token_idx,
+                input_amount,
+            } => {
+                if input_token_idx == output_token_idx {
+                    return Err(AmmError::EqualTokensError);
+                }
+
+                let input_pool_balance = *self.get_pool_balance(input_token_idx)?;
+                let output_pool_balance = *self.get_pool_balance(output_token_idx)?;
+
+                let output_amount = amm::calculate_output_amount(
+                    input_amount,
+                    input_pool_balance,
+                    output_pool_balance,
+                )?;
+
+                self.decrease_pool_balance(input_token_idx, input_amount)?;
+                self.increase_pool_balance(output_token_idx, output_amount)?;
+
+                Ok(ExecutionResult::default())
+            }
+            Self::Operation::AddLiquidity {
+                token0_amount,
+                token1_amount,
+            } => {
+                let balance0 = self.balance0.get_mut();
+                let balance1 = self.balance1.get_mut();
+
+                let balance0_float = *balance0 as f64;
+                let balance1_float = *balance1 as f64;
+
+                let token0_amount_float = token0_amount as f64;
+                let token1_amount_float = token1_amount as f64;
+
+                let total_shares = self.total_shares.get_mut();
+                let total_shares_float = *total_shares as f64;
+
+                if *total_shares > 0
+                    && balance0_float / balance1_float != token0_amount_float / token1_amount_float
+                {
+                    return Err(AmmError::BalanceRatioAlteredError);
+                }
+
+                if *total_shares > 0
+                    && (token0_amount_float / balance0_float) * total_shares_float
+                        != (token1_amount_float / balance1_float) * total_shares_float
+                {
+                    return Err(AmmError::TotalSharesRatioAlteredError);
+                }
+
+                let new_shares = if *total_shares == 0 {
+                    (token0_amount_float * token1_amount_float).sqrt() as u64
+                } else {
+                    ((token0_amount_float / balance0_float) * total_shares_float) as u64
+                };
+
+                *balance0 += token0_amount;
+                *balance1 += token1_amount;
+
+                *total_shares += new_shares;
```

Comment:
In my current version (using fungible token), I still store this `total_shares`. Would I not need to? Can I always calculate the total shares with `sqrt(token0_amount * token1_amount)`? I'll update this PR in a second, you can take a look and see the current code for this, to help answer this question

Reply:
I'm actually not sure why you need the geometric mean at all, and not just the product.

And for the product, I'd say it's easier to always compute it when needed, yes. It can't stay _exactly_ constant anyway, so I think the rule should just be that it never decreases.

Reply:
I see, so remove the concept of shares completely then, right? That was also from the explanation example I learned from haha didn't know it could be removed! Let me work on that then, before the other stuff